### ISSUE TITLE
Machine readable output

### ISSUE THREAD

**Issue by kadel** [openshift/odo#784]
Add option to specify the output format.
For example  `odo -o json` will make ODO output formatted as JSON (all tables in `list` commands will be displayed in json format etc..)
Each command needs to have output JSON structure documented in our documentation.

The goal of this is to make it easier for 3rd party programs (like  [vscode-openshift-tools](https://github.com/redhat-developer/vscode-openshift-tools) to parse odo output.

- [x]  Create structs for machine readable output #1062
- [ ] Followup on machine readable output (PR#1196) #1263

**Comment by cmoulliard** [comment]
We should offer 2 output format at least -> json and yaml as yaml is more appropriated for terminal display. Make sense ?

**Comment by kadel** [comment]
> We should offer 2 output format at least -> json and yaml as yaml is more appropriated for terminal display. Make sense ?

Yep, makes sense.

**Comment by jorgemoralespou** [comment]
What would be the default output name? There should also need to be a `-o` option for it, to be specific about the format to use.

**Comment by kadel** [comment]
> What would be the default output name? There should also need to be a `-o` option for it, to be specific about the format to use.

Can't think of a good name :-( How about just `-o text`?

**Comment by cdrage** [comment]
@kadel @jorgemoralespou @cmoulliard 

Am I correct in saying that we are blocked on https://github.com/redhat-developer/odo/issues/785 before proceeding / implementing the machine-readable output for Odo?

**Comment by kadel** [comment]
Would it make sense to use Kubernets objects syntax for machine-readable output?

Examples:
```
odo describe my-spring-boot
```
```
apiVersion: component.k8s.io/v1alpha1
kind: Component
metadata:
  name: my-spring-boot
spec:
  type: springboot  
  version: 1.5.16
  source: file:///Users/tomas/tmp/odo-examples/mysqlpython
  envs:
  - name: SPRING_PROFILES_ACTIVE
    value: openshift-catalog
```


`odo create` if successful would also return the description of the component that was created
`odo list` would return list of component objects


What do you think @jorgemoralespou @cmoulliard @sspeiche?
@dgolovin is this something that you could work with?


How would be handle errors? It would be nice to return a bit more information than just error exit code.

**Comment by jorgemoralespou** [comment]
@kadel I would think this more of a `--dry-run` command rather than the machine readable output.

For machine readable output I think we would need to provide the same type of output for every action but in different format, which means, that if you currently get as output of `odo create ...` something like this:

```
$ odo create nodejs backend
Component 'backend' was created and port 8080/TCP was opened
To push source code to the component run 'odo push'

Component 'backend' is now set as active component.
```

In yaml output would be:
```
odo create nodejs backend -o yaml
component: backend
result: success
ports:
  - 8080/tcp
```

Or something similar :-D

**Comment by kadel** [comment]
> In yaml output would be:
> 
> ```
> odo create nodejs backend -o yaml
> component: backend
> result: success
> ports:
>   - 8080/tcp
> ```
> Or something similar :-D


Yep, that was my idea, the only difference is that I would omit `result: success`, as you haave this information also in form of exit code. And adding more info to the yaml structure, to contain full info about the component that was just created. I was thinking about using Kubernetes resource format because it has standard basic strucutre and versioning.

**Comment by surajnarwade** [comment]
while working on https://github.com/redhat-developer/odo/issues/853, I was thinking about this issue as well, so I modified as follows:

* Output of `odo describe` which can be modified easily:

```
$ odo describe httpd-ex-httpd-qgxb
Name: httpd-ex-httpd-qgxb
Type: httpd
Source: file:///home/snarwade/odo-example/httpd-ex
Env:
 - a: b
 - d: v
Storage:
 - data1: 100M
```

* `component describe` struct looks like this:

```
type ComponentDesc struct {
	ComponentImageType string
	Path               string
	ComponentURL       string
	ComponentPort      int
	ComponentProtocol  string
	Env                []corev1.EnvVar
	AppStore           []storage.StorageInfo
}
```


* for machine readable output, we can just marshal this struct into json and we can do something like:

```
odo describe httpd-ex-httpd-qgxb -o json
{
  "Name": "httpd-ex-httpd-qgxb",
  "Type": "httpd",
  "Source": "file:///home/snarwade/odo-example/httpd-ex",
  "Env": [
    {
      "a": "b"
    },
    {
      "d": "v"
    }
  ],
  "Storage": [
    {
      "data1": "100M"
    }
  ]
}
```

WDYT @kadel  @jorgemoralespou @cmoulliard @cdrage  ?

**Comment by jorgemoralespou** [comment]
@surajnarwade to me looks ok, but only one thing. I would not have the result in Capitalized, but rather camelCase, as I think that seems to be the standard.

e.g.
```
- apiVersion: v1
  kind: Pod
  metadata:
    annotations:
    creationTimestamp:
    labels:
    name:
    namespace:
    ownerReferences:
```

```
            "apiVersion": "v1",
            "kind": "Pod",
            "metadata": {
                "annotations": {
                },
                "creationTimestamp": "",
                "labels": {
                },
                "name": "",
                "namespace": "",
                "ownerReferences": []
```

**Comment by surajnarwade** [comment]
@jorgemoralespou , this was just POC, I will add camelCasing. I am wondering, is it okay to have default output as YAML (as mentioned)

**Comment by jorgemoralespou** [comment]
@surajnarwade what do you mean by default as yaml? Default should be `text`. What we have agreed is human readable text. These other machine readable outputs should only happen if you provide a `-o` (or whatever is defined) flag.

**Comment by kadel** [comment]
> @jorgemoralespou , this was just POC, I will add camelCasing. I am wondering, is it okay to have default output as YAML (as mentioned)

Default output should be something that is human readable, and it is easy to "parse" by just looking at it.

It can be something that resembles yaml. But it shouldn't be strictly yaml. Good visualization is the key factor.

**Comment by kadel** [comment]
@surajnarwade 

Your examples look a bit too verbose for "normal" output

for example here:
```
$ odo describe httpd-ex-httpd-qgxb
Component Name: httpd-ex-httpd-qgxb
Type: httpd
Source: file:///home/snarwade/odo-example/httpd-ex
Environment:
- name: env1
  value: value1
- name: env2
  value: value2
Storage:
- Name: data1
  Path: /data
  Size: 100M
URL:
- Name: httpd-ex-httpd-qgxb
  Port: 8080
  Protocol: http
  URL: httpd-ex-httpd-qgxb-httpd-ex-tifo-myproject.192.168.42.65.nip.io
```


I think that somehting like this is much better as normal output
```
$ odo describe httpd-ex-httpd-qgxb

Name: httpd-ex-httpd-qgxb
Type: httpd
Source: file:///home/snarwade/odo-example/httpd-ex
Environment variables:
  - env1=value1
  - env2=value2
Storage:
 - mystorage mounted to /data
URLs:
  - httpd-ex-httpd-qgxb-httpd-ex-tifo-myproject.192.168.42.65.nip.io:8080
```

You don't have to dump all the information for storage,urls etc..... just important bits There are other commands to do that.

**Comment by surajnarwade** [comment]
cool, I agree to the point @kadel @jorgemoralespou , thanks for the feedback

**Comment by girishramnani** [comment]
we can also consider a `-filter "{{ .URLs }}"` similar to json paths in kuberenetes

**Comment by dgolovin** [comment]
@kadel noticed that some information is missing when I do `odo describe` for component
```json
{
    "kind": "Component",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "name": "nodjslatestodo",
        "creationTimestamp": null
    },
    "spec": {
        "type": "nodejs",
        "source": "https://github.com/dgolovin/nodejs-ex",
        "url": [
            "nodjslatestodo-8080"
        ]
    },
    "status": {
        "active": true
    }
}
```

* creationTimestamp is null
* type should also have version included `nodejs:latest` or as an additional field `version`

Also `url` references url by name, which would require to use something like `odo url describe` to fetch additional info, but it is not available.

**Comment by surajnarwade** [comment]
Hi @dgolovin ,

* `creationTimeStamp` is not in our control, and I don't think we need to update this field as of now
* good point about the tag of image, :+1:

**Comment by dgolovin** [comment]
@kadel @sudhirverma noticed `odo project list -o json` is not supported yet in odo 0.0.19.

**Comment by girishramnani** [comment]
this issue is superseded by https://github.com/openshift/odo/issues/1739

**Comment by girishramnani** [comment]
closing
